wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/FactoryManager.java

```diff
@@ -0,0 +1,283 @@
+     *
+     * @param factoryId
+     *         factory identifier
+     * @param imageId
+     *         image identifier
+     * @return factory images or empty set if no image found by given {@code imageId}
+     * @throws NotFoundException
+     *         when specified factory not found
+     * @throws ServerException
+     *         when any server errors occurs
+     */
+    public Set<FactoryImage> getFactoryImages(String factoryId, String imageId) throws NotFoundException,
+                                                                                       ServerException {
+        requireNonNull(factoryId);
+        requireNonNull(imageId);
+        return getFactoryImages(factoryId).stream()
+                                          .filter(image -> imageId.equals(image.getName()))
+                                          .collect(Collectors.toSet());
+    }
+
+    /**
+     * Gets all the factory images.
+     *
+     * @param factoryId
+     *         factory identifier
+     * @return factory images or empty set if no image found for factory
+     * @throws NotFoundException
+     *         when specified factory not found
+     * @throws ServerException
+     *         when any server errors occurs
+     */
+    public Set<FactoryImage> getFactoryImages(String factoryId) throws NotFoundException,
+                                                                       ServerException {
+        requireNonNull(factoryId);
+        return factoryDao.getById(factoryId).getImages();
+    }
+
+    /**
+     * Get list of factories which conform specified attributes.
+     *
+     * @param maxItems
+     *         max number of items in response
+     * @param skipCount
+     *         skip items. Must be equals or greater then {@code 0}
+     * @param attributes
+     *         skip items. Must be equals or greater then {@code 0}
+     * @return stored data, if specified attributes is correct
+     * @throws ServerException
+     *         when any server errors occurs
+     */
+    public List<? extends Factory> getByAttribute(int maxItems,
+                                                  int skipCount,
+                                                  List<Pair<String, String>> attributes) throws ServerException {
+        return factoryDao.getByAttribute(maxItems, skipCount, attributes);
+    }
+
+    /**
+     * Gets factory snippet by factory id and snippet type.
+     * If snippet type is not set, "url" type will be used as default.
+     *
+     * @param factoryId
+     *         id of factory
+     * @param snippetType
+     *         type of snippet
+     * @param uriInfo
+     *         url context
+     * @return snippet content.
+     * @throws NotFoundException
+     *         when factory with specified id doesn't not found - with response code 400
+     * @throws ServerException
+     *         when any server error occurs during snippet creation
+     */
+    public String getFactorySnippet(String factoryId,
```

Comment:
Mark the method with `@Nullable` or factory doesn't follow _everything is not-null by default_ pattern ?